---
title: Arrêt des erreurs | Microsoft Docs
ms.custom: ''
ms.date: 09/13/2016
ms.reviewer: ''
ms.suite: ''
ms.tgt_pltfrm: ''
ms.topic: article
ms.assetid: b804e738-aefa-41bb-9649-f9ed897fd96c
caps.latest.revision: 8
ms.openlocfilehash: d1967fe7996f75ec5229920f7ec49aa5ff6bdbfd
ms.sourcegitcommit: caac7d098a448232304c9d6728e7340ec7517a71
ms.translationtype: MT
ms.contentlocale: fr-FR
ms.lasthandoff: 03/16/2019
ms.locfileid: "58059233"
---
# <a name="terminating-errors"></a><span data-ttu-id="411d3-102">Erreur avec fin d’exécution</span><span class="sxs-lookup"><span data-stu-id="411d3-102">Terminating Errors</span></span>

<span data-ttu-id="411d3-103">Cette rubrique décrit la méthode utilisée pour signaler les erreurs de fin d’exécution.</span><span class="sxs-lookup"><span data-stu-id="411d3-103">This topic discusses the method used to report terminating errors.</span></span> <span data-ttu-id="411d3-104">Elle explique également comment appeler la méthode à partir de l’applet de commande, et il traite les exceptions qui peuvent être retournées par le runtime de Windows PowerShell lorsque la méthode est appelée.</span><span class="sxs-lookup"><span data-stu-id="411d3-104">It also discusses how to call the method from within the cmdlet, and it discusses the exceptions that can be returned by the Windows PowerShell runtime when the method is called.</span></span>

<span data-ttu-id="411d3-105">Lorsqu’une fin erreur se produit, l’applet de commande doit signaler l’erreur en appelant le [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) (méthode).</span><span class="sxs-lookup"><span data-stu-id="411d3-105">When a terminating error occurs, the cmdlet should report the error by calling the [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) method.</span></span> <span data-ttu-id="411d3-106">Cette méthode permet à l’applet de commande Envoyer un enregistrement d’erreur qui décrit la condition ayant provoqué l’erreur avec fin d’exécution.</span><span class="sxs-lookup"><span data-stu-id="411d3-106">This method allows the cmdlet to send an error record that describes the condition that caused the terminating error.</span></span> <span data-ttu-id="411d3-107">Pour plus d’informations sur les enregistrements d’erreur, consultez [enregistrements d’erreur Windows PowerShell](./windows-powershell-error-records.md).</span><span class="sxs-lookup"><span data-stu-id="411d3-107">For more information about error records, see [Windows PowerShell Error Records](./windows-powershell-error-records.md).</span></span>

<span data-ttu-id="411d3-108">Lorsque le [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) est appelée, le runtime Windows PowerShell arrête l’exécution du pipeline définitivement et lève un [ System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception.</span><span class="sxs-lookup"><span data-stu-id="411d3-108">When the [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) method is called, the  Windows PowerShell runtime permanently stops the execution of the pipeline and throws a [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception.</span></span> <span data-ttu-id="411d3-109">Toute tentative ultérieure de l’appeler [System.Management.Automation.Cmdlet.WriteObject](/dotnet/api/System.Management.Automation.Cmdlet.WriteObject), [System.Management.Automation.Cmdlet.WriteError](/dotnet/api/System.Management.Automation.Cmdlet.WriteError), ou plusieurs autres API provoque ces appels lève une [ System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception.</span><span class="sxs-lookup"><span data-stu-id="411d3-109">Any subsequent attempts to call [System.Management.Automation.Cmdlet.WriteObject](/dotnet/api/System.Management.Automation.Cmdlet.WriteObject), [System.Management.Automation.Cmdlet.WriteError](/dotnet/api/System.Management.Automation.Cmdlet.WriteError), or several other APIs causes those calls to throw a [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception.</span></span>

<span data-ttu-id="411d3-110">Le [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception peut également se produire si une autre applet de commande dans le pipeline signale une erreur avec fin d’exécution, si l’utilisateur a demandé d’arrêter le pipeline, ou si le pipeline a été arrêté. avant la fin pour une raison quelconque.</span><span class="sxs-lookup"><span data-stu-id="411d3-110">The [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception can also occur if another cmdlet in the pipeline reports a terminating error, if the user has asked to stop the pipeline, or if the pipeline has been halted before completion for any reason.</span></span> <span data-ttu-id="411d3-111">L’applet de commande n’a pas besoin d’intercepter le [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception, sauf si elle doit ouvrir Nettoyage des ressources ou son état interne.</span><span class="sxs-lookup"><span data-stu-id="411d3-111">The cmdlet does not need to catch the [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception unless it must clean up open resources or its internal state.</span></span>

<span data-ttu-id="411d3-112">Applets de commande peut écrire n’importe quel nombre d’objets de sortie ou des erreurs sans fin d’exécution avant de signaler une erreur avec fin d’exécution.</span><span class="sxs-lookup"><span data-stu-id="411d3-112">Cmdlets can write any number of output objects or non-terminating errors before reporting a terminating error.</span></span> <span data-ttu-id="411d3-113">Toutefois, l’erreur avec fin d’exécution s’arrête définitivement le pipeline et aucune sortie supplémentaire, fin d’exécution, les erreurs ou les erreurs sans fin d’exécution peuvent être signalées.</span><span class="sxs-lookup"><span data-stu-id="411d3-113">However, the terminating error permanently stops the pipeline, and no further output, terminating errors, or non-terminating errors can be reported.</span></span>

<span data-ttu-id="411d3-114">Applets de commande peut appeler [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) uniquement à partir du thread qui a appelé le [System.Management.Automation.Cmdlet.BeginProcessing](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing), [ System.Management.Automation.Cmdlet.ProcessRecord](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord), ou [System.Management.Automation.Cmdlet.EndProcessing](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing) méthode de traitement d’entrée.</span><span class="sxs-lookup"><span data-stu-id="411d3-114">Cmdlets can call [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) only from the thread that called the [System.Management.Automation.Cmdlet.BeginProcessing](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing), [System.Management.Automation.Cmdlet.ProcessRecord](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord), or [System.Management.Automation.Cmdlet.EndProcessing](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing) input processing method.</span></span> <span data-ttu-id="411d3-115">N’essayez pas d’appeler [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) ou [System.Management.Automation.Cmdlet.WriteError](/dotnet/api/System.Management.Automation.Cmdlet.WriteError) à partir d’un autre thread.</span><span class="sxs-lookup"><span data-stu-id="411d3-115">Do not attempt to call [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) or [System.Management.Automation.Cmdlet.WriteError](/dotnet/api/System.Management.Automation.Cmdlet.WriteError) from another thread.</span></span> <span data-ttu-id="411d3-116">Au lieu de cela, les erreurs doivent être communiquées au thread principal.</span><span class="sxs-lookup"><span data-stu-id="411d3-116">Instead, errors must be communicated back to the main thread.</span></span>

<span data-ttu-id="411d3-117">Il est possible pour une applet de commande pour lever une exception dans son implémentation de la [System.Management.Automation.Cmdlet.BeginProcessing](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing), [System.Management.Automation.Cmdlet.ProcessRecord](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord), ou [System.Management.Automation.Cmdlet.EndProcessing](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing) (méthode).</span><span class="sxs-lookup"><span data-stu-id="411d3-117">It is possible for a cmdlet to throw an exception in its implementation of the [System.Management.Automation.Cmdlet.BeginProcessing](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing), [System.Management.Automation.Cmdlet.ProcessRecord](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord), or [System.Management.Automation.Cmdlet.EndProcessing](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing) method.</span></span> <span data-ttu-id="411d3-118">Toute exception levée à partir de ces méthodes (à l’exception de certaines conditions d’erreur grave qui arrêter l’hôte Windows PowerShell) est interprétée comme une erreur avec fin d’exécution qui arrête le pipeline, mais pas Windows PowerShell dans sa globalité.</span><span class="sxs-lookup"><span data-stu-id="411d3-118">Any exception thrown from these methods (except for a few severe error conditions that stop the Windows PowerShell host) is interpreted as a terminating error which stops the pipeline, but not Windows PowerShell as a whole.</span></span> <span data-ttu-id="411d3-119">(Cela s’applique uniquement au thread principal applet de commande.</span><span class="sxs-lookup"><span data-stu-id="411d3-119">(This applies only to the main cmdlet thread.</span></span> <span data-ttu-id="411d3-120">Les exceptions non interceptées dans les threads engendrés par l’applet de commande, en général, l’arrêter l’hôte Windows PowerShell.) Nous vous recommandons d’utiliser [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) au lieu de lever une exception, car l’enregistrement d’erreur fournit des informations supplémentaires sur la condition d’erreur, qui est utile pour l’utilisateur final.</span><span class="sxs-lookup"><span data-stu-id="411d3-120">Uncaught exceptions in threads spawned by the cmdlet, in general, halt the Windows PowerShell host.) We recommend that you use [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) rather than throwing an exception because the error record provides additional information about the error condition, which is useful to the end-user.</span></span> <span data-ttu-id="411d3-121">Applets de commande doivent respecter les instructions de code managé par rapport à la mise en cache et la gestion des exceptions de tous les (`catch (Exception e)`).</span><span class="sxs-lookup"><span data-stu-id="411d3-121">Cmdlets should honor the managed code guideline against catching and handling all exceptions (`catch (Exception e)`).</span></span> <span data-ttu-id="411d3-122">Seules les exceptions de types connus et attendus convertissent des enregistrements d’erreur.</span><span class="sxs-lookup"><span data-stu-id="411d3-122">Convert only exceptions of known and expected types into error records.</span></span>

## <a name="see-also"></a><span data-ttu-id="411d3-123">Voir aussi</span><span class="sxs-lookup"><span data-stu-id="411d3-123">See Also</span></span>

[<span data-ttu-id="411d3-124">System.Management.Automation.Cmdlet.BeginProcessing</span><span class="sxs-lookup"><span data-stu-id="411d3-124">System.Management.Automation.Cmdlet.BeginProcessing</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing)

[<span data-ttu-id="411d3-125">System.Management.Automation.Cmdlet.EndProcessing</span><span class="sxs-lookup"><span data-stu-id="411d3-125">System.Management.Automation.Cmdlet.EndProcessing</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing)

[<span data-ttu-id="411d3-126">System.Management.Automation.Cmdlet.ProcessRecord</span><span class="sxs-lookup"><span data-stu-id="411d3-126">System.Management.Automation.Cmdlet.ProcessRecord</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord)

[<span data-ttu-id="411d3-127">System.Management.Automation.Pipelinestoppedexception</span><span class="sxs-lookup"><span data-stu-id="411d3-127">System.Management.Automation.Pipelinestoppedexception</span></span>](/dotnet/api/System.Management.Automation.PipelineStoppedException)

[<span data-ttu-id="411d3-128">System.Management.Automation.Cmdlet.Throwterminatingerror\*</span><span class="sxs-lookup"><span data-stu-id="411d3-128">System.Management.Automation.Cmdlet.Throwterminatingerror\*</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError)

[<span data-ttu-id="411d3-129">System.Management.Automation.Cmdlet.WriteError</span><span class="sxs-lookup"><span data-stu-id="411d3-129">System.Management.Automation.Cmdlet.WriteError</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.WriteError)

[<span data-ttu-id="411d3-130">Windows PowerShell erreur enregistrements</span><span class="sxs-lookup"><span data-stu-id="411d3-130">Windows PowerShell Error Records</span></span>](./windows-powershell-error-records.md)

[<span data-ttu-id="411d3-131">Écriture d’une applet de commande Windows PowerShell</span><span class="sxs-lookup"><span data-stu-id="411d3-131">Writing a Windows PowerShell Cmdlet</span></span>](./writing-a-windows-powershell-cmdlet.md)
